File: javascript/src/sign.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const { parseArgs } = require('node:util');
const { PDFDocument } = require('pdf-lib');
const signpdf = require('@signpdf/signpdf').default;
const { pdflibAddPlaceholder } = require('@signpdf/placeholder-pdf-lib');
const { PemSigner } = require('./pemSigner');

/**
 * Sign `inputPath` and write the result to `outputPath`.
 */
async function signPdf({
  inputPath,
  outputPath,
  keyPath,
  certPath,
  chainPaths = [],
  passphrase = null,
  reason = '',
  location = '',
  contactInfo = '',
  signerName = '',
}) {
  const privateKeyPem = fs.readFileSync(keyPath);
  const certificatePem = fs.readFileSync(certPath);
  const caChainPems = chainPaths.map((p) => fs.readFileSync(p));

  const signer = new PemSigner(privateKeyPem, certificatePem, { caChainPems, passphrase });

  const pdfDoc = await PDFDocument.load(fs.readFileSync(inputPath));
  pdflibAddPlaceholder({
    pdfDoc,
    reason,
    contactInfo,
    name: signerName,
    location,
  });
  const pdfWithPlaceholder = await pdfDoc.save({ useObjectStreams: false });

  const signedPdf = await signpdf.sign(Buffer.from(pdfWithPlaceholder), signer);
  fs.writeFileSync(outputPath, signedPdf);
}

function parseCliArgs(argv) {
  const { values, positionals } = parseArgs({
    args: argv,
    allowPositionals: true,
    options: {
      key: { type: 'string' },
      cert: { type: 'string' },
      chain: { type: 'string', multiple: true, default: [] },
      passphrase: { type: 'string' },
      reason: { type: 'string', default: '' },
      location: { type: 'string', default: '' },
      'contact-info': { type: 'string', default: '' },
      'signer-name': { type: 'string', default: '' },
    },
  });

  if (positionals.length !== 2 || !values.key || !values.cert) {
    console.error(
      'Usage: sign-pdf <input.pdf> <output.pdf> --key key.pem --cert cert.pem ' +
        '[--chain CERT...] [--passphrase PASSPHRASE] [--reason REASON] ' +
        '[--location LOCATION] [--contact-info INFO] [--signer-name NAME]',
    );
    process.exit(1);
  }

  const [inputPath, outputPath] = positionals;
  return {
    inputPath,
    outputPath,
    keyPath: values.key,
    certPath: values.cert,
    chainPaths: values.chain,
    passphrase: values.passphrase ?? null,
    reason: values.reason,
    location: values.location,
    contactInfo: values['contact-info'],
    signerName: values['signer-name'],
  };
}

async function main() {
  const args = parseCliArgs(process.argv.slice(2));
  await signPdf(args);
  console.log(`Signed PDF written to ${args.outputPath}`);
}

if (require.main === module) {
  main().catch((err) => {
    console.error(err.message);
    process.exit(1);
  });
}

module.exports = { signPdf };

File: src/websignpdf/sign.py
"""Sign a PDF file with a private key and X.509 certificate.

Wraps pyHanko to add a digital signature field to an existing PDF, signed
with a PEM-encoded private key and certificate (plus an optional chain of
intermediate certificates).
"""

from __future__ import annotations

import argparse
import sys

from pyhanko.pdf_utils.incremental_writer import IncrementalPdfFileWriter
from pyhanko.sign import fields, signers


def sign_pdf(
    input_path: str,
    output_path: str,
    key_path: str,
    cert_path: str,
    chain_paths: list[str] | None = None,
    passphrase: bytes | None = None,
    field_name: str = "Signature1",
    reason: str | None = None,
    location: str | None = None,
    contact_info: str | None = None,
    signer_name: str | None = None,
) -> None:
    """Sign ``input_path`` and write the result to ``output_path``."""

    signer = signers.SimpleSigner.load(
        key_file=key_path,
        cert_file=cert_path,
        ca_chain_files=chain_paths,
        key_passphrase=passphrase,
    )
    if signer is None:
        raise ValueError(
            f"Could not load private key/certificate from {key_path!r} / {cert_path!r}"
        )

    signature_meta = signers.PdfSignatureMetadata(
        field_name=field_name,
        reason=reason,
        location=location,
        contact_info=contact_info,
        name=signer_name,
    )

    with open(input_path, "rb") as inf:
        writer = IncrementalPdfFileWriter(inf)
        out = signers.sign_pdf(
            writer,
            signature_meta,
            signer=signer,
            new_field_spec=fields.SigFieldSpec(sig_field_name=field_name),
        )

    with open(output_path, "wb") as outf:
        outf.write(out.getvalue())


def main(argv: list[str] | None = None) -> None:
    parser = argparse.ArgumentParser(description="Sign a PDF file with a private key and certificate.")
    parser.add_argument("input", help="Path to the input PDF file")
    parser.add_argument("output", help="Path to write the signed PDF file")
    parser.add_argument("--key", required=True, help="Path to the PEM-encoded private key")
    parser.add_argument("--cert", required=True, help="Path to the PEM-encoded signing certificate")
    parser.add_argument(
        "--chain",
        nargs="*",
        default=None,
        metavar="CERT",
        help="Additional PEM-encoded certificates to include in the signature chain",
    )
    parser.add_argument("--passphrase", default=None, help="Passphrase for the private key, if encrypted")
    parser.add_argument("--field-name", default="Signature1", help="Name of the signature field to create")
    parser.add_argument("--reason", default=None, help="Reason for signing, shown to the viewer")
    parser.add_argument("--location", default=None, help="Location of signing, shown to the viewer")
    parser.add_argument("--contact-info", default=None, help="Signer's contact information")
    parser.add_argument("--signer-name", default=None, help="Display name of the signer")
    args = parser.parse_args(argv)

    passphrase = args.passphrase.encode("utf-8") if args.passphrase else None

    sign_pdf(
        input_path=args.input,
        output_path=args.output,
        key_path=args.key,
        cert_path=args.cert,
        chain_paths=args.chain,
        passphrase=passphrase,
        field_name=args.field_name,
        reason=args.reason,
        location=args.location,
        contact_info=args.contact_info,
        signer_name=args.signer_name,
    )
    print(f"Signed PDF written to {args.output}")


if __name__ == "__main__":
    sys.exit(main())
